engine/gui/widgets.py

```diff
@@ -84,15 +86,85 @@ def create(
         selected_button_name = spec.get("initial_selected_button")
         selected_button = None
         if selected_button_name is not None:
-            selected_button = next(
-                (b for b in buttons if b.name == selected_button_name),
-                None,
-            )
-
-        return GUISpec(
+            try:
+                selected_button = next(
+                    (b for b in buttons if b.name == selected_button_name)
+                )
+            except StopIteration as exc:
+                raise ValidationError(
+                    f"Initial select button {selected_button_name} does not exist."
+                ) from exc
+
+        gui_spec = GUISpec(
             assets=[Asset(**a) for a in spec.get("assets", [])],
             buttons=buttons,
             images=[Image(**b) for b in spec.get("images", [])],
             cancel_action=cancel_action,
             initial_selected_button=selected_button,
         )
+
+        gui_spec.validate()
+        return gui_spec
+
+    def validate(self) -> None:
+        """Validates the GUISpec."""
+
+        if not self.buttons and not self.images:
+            raise ValidationError("Must have at least one button or image")
+
+        asset_names = collections.Counter([a.name for a in self.assets])
+        duplicate_assets = [asset for asset, count in asset_names.items() if count > 1]
+        if duplicate_assets:
+            raise ValidationError(f"Duplicate asset name: {duplicate_assets}.")
+
+        button_names = collections.Counter([b.name for b in self.buttons])
+        duplicate_buttons = [name for name, count in button_names.items() if count > 1]
+        if duplicate_buttons:
+            raise ValidationError(f"Duplicate button name: {duplicate_buttons}")
+
+        image_names = collections.Counter([i.image_asset for i in self.images])
+        duplicate_images = [image for image, count in image_names.items() if count > 1]
+        if duplicate_images:
+            raise ValidationError(f"Duplicate image name(s): {duplicate_images}")
+
+        for button in self.buttons:
+            if button.selected_image_asset not in [a.name for a in self.assets]:
```

Comment:
I'll merge it but something like this is inefficient:
```python
for button in self.buttons:
    if button.selected_image_asset not in [a.name for a in self.assets]
```
The reason is that to calculate `[a.name for a in self.assets]` it takes O(n) steps, and then doing it inside a loop makes it O(n^2). It doesn't really matter here, but since this is a bad habit that new developers have you should break it as soon as possible or you'll be churning out lots of low-performance code :)

A better approach is to just use the counter that already exists:
```python
for button in self.buttons:
    if button.selected_image_asset not in asset_names:
```
Doing a `not in` on a `Counter` (which is a subclass of `dict`) is a O(1) operation, so in a loop it will be O(n).

In the case that the counter didn't already exist, it'd be best to just construct a set:
```python
asset_names = {a.name for a in self.assets}
```